### PR TITLE
ci: add github action workflows

### PR DESCRIPTION
**Pull request by laojianzi**
Trying to solve: #2582 

Merge before: 

- open [Setting > Secrets](https://github.com/gin-gonic/gin/settings/secrets/actions) and set you'r gitter room id to `GITTER_ROOM_ID `
- Merge #2595

**If the master branch changes, or if a problem occurs that does not necessarily occur, you can choose to re-run the github action job.**

Differences from Travis CI:

- Multiple os `ubuntu-latest`, `macos-latest`, `windows-latest`
- No go master version
- Sending gitter notifications using curl

---

### Some of the problems I've encountered since using GitHub Action

- `TestUnixSocket` windows cannot read `/tmp/unix_unit_test` **(fixed: #2595)**
- `TestPathCleanMallocs` have a certain chance of failure **(I don't know how to repair it :confused: )**
>     path_test.go:85: 
>        	Error Trace:	path_test.go:85
>        	Error:      	Not equal: 
>        	            	expected: float64(2010)
>        	            	actual  : int(0)
>        	Test:       	TestPathCleanMallocs
- `TestUnixSocket` have a certain chance of failure **(I don't know how to repair it :confused: )**
> Get http://127.0.0.1:49379/test: dial tcp 127.0.0.1:49379: bad file descriptor

### CODE REVIEW

**Comment by thinkerou** [comment]
`open Setting > Secrets and set you'r gitter room id to GITTER_ROOM_ID`
-- now only @javierprovecho can set it

**Comment by laojianzi** [comment]
> `open Setting > Secrets and set you'r gitter room id to GITTER_ROOM_ID`
> -- now only @javierprovecho can set it

https://github.com/gin-gonic/gin/blob/b01605bb5b43dbf33781970af5ad6633e5549fd1/.travis.yml#L47

@thinkerou Use `7f95bf605c4d356372f4`, as in `.travis.yml`.

What do you think?

**Comment by thinkerou** [comment]
> > `open Setting > Secrets and set you'r gitter room id to GITTER_ROOM_ID`
> > -- now only @javierprovecho can set it
> 
> https://github.com/gin-gonic/gin/blob/b01605bb5b43dbf33781970af5ad6633e5549fd1/.travis.yml#L47
> 
> @thinkerou Replace `GITTER_ROOM_ID` with `7f95bf605c4d356372f4`, as in `.travis.yml`.
> 
> What do you think?

Yes, OK! thanks.

File: .github/workflows/gin.yml
Comment (thinkerou) [inline .github/workflows/gin.yml:48]: why `matrix.go == 1.15` here?

Comment (laojianzi) [reply .github/workflows/gin.yml:48]: If the `go test` is successful, the coverage is theoretically the same every time

So here we just need to avoid multiple uploads and choose one of them

Comment (thinkerou) [reply .github/workflows/gin.yml:48]: @laojianzi ok, thank your reply.

Comment (laojianzi) [reply .github/workflows/gin.yml:48]: 1.12.x: https://travis-ci.org/github/gin-gonic/gin/jobs/753866331#L1150
1.13.x: https://travis-ci.org/github/gin-gonic/gin/jobs/753866332#L1150
...

@thinkerou In Travis CI though, it seems that the go test coverage is uploaded for every go version, so I can update it if you want.

Comment (thinkerou) [reply .github/workflows/gin.yml:48]: @laojianzi yes, please update, and remove `.travis.yml` and update gitter-room-id,thanks

Comment (laojianzi) [reply .github/workflows/gin.yml]: @thinkerou updated

**Comment by laojianzi** [comment]
https://github.com/gin-gonic/gin/runs/1706106769?check_suite_focus=true#step:6:32
```
--- FAIL: TestLoadHTMLFilesDebugMode (0.00s)
make: *** [test] Error 1
    debug_test.go:144: 
        	Error Trace:	debug_test.go:144
        	            				asm_amd64.s:1357
        	Error:      	Received unexpected error:
        	            	read |0: bad file descriptor
        	Test:       	TestLoadHTMLFilesDebugMode
```

`bad file descriptor` often occurs, suspected to be related to the use of `goroutine` in testing (speculation without any evidence)

**Review by thinkerou — DISMISSED** [review]
lgtm, thanks!

**Comment by thinkerou** [comment]
@appleboy please review the pr, thanks!

`continuous-integration/travis-ci Expected — Waiting for status to be reported`
why still required?

**Comment by laojianzi** [comment]
@thinkerou Maybe it has something to do with this

image

**Comment by laojianzi** [comment]
https://github.com/gin-gonic/gin/blob/540b1eff7069128df5d95e09968569f2266d9a6a/.github/PULL_REQUEST_TEMPLATE.md#L4
https://github.com/gin-gonic/gin/blob/540b1eff7069128df5d95e09968569f2266d9a6a/CONTRIBUTING.md#L11
https://github.com/gin-gonic/gin/blob/46ddd4259cac975be1eb11b4f1192264f582db16/README.md#L5

@thinkerou I think these need to be replaced with GitHub Action, do they need to be updated here?

**Comment by thinkerou** [comment]
> https://github.com/gin-gonic/gin/blob/540b1eff7069128df5d95e09968569f2266d9a6a/.github/PULL_REQUEST_TEMPLATE.md#L4
> 
> 
> https://github.com/gin-gonic/gin/blob/540b1eff7069128df5d95e09968569f2266d9a6a/CONTRIBUTING.md#L11
> 
> 
> https://github.com/gin-gonic/gin/blob/46ddd4259cac975be1eb11b4f1192264f582db16/README.md#L5
> 
> @thinkerou I think these need to be replaced with GitHub Action, do they need to be updated here?

yes, please update, thanks!

**Comment by laojianzi** [comment]
@thinkerou @appleboy okay, plz review, thanks!

**Comment by thinkerou** [comment]
> @thinkerou Maybe it has something to do with this, only @javierprovecho can set it?
> 
> image

yes, please @javierprovecho review and set, thanks!

File: .github/workflows/gin.yml
Comment (appleboy) [inline .github/workflows/gin.yml:16]: also add `1.16`

**Review by appleboy — CHANGES_REQUESTED** [review]

**Comment by appleboy** [comment]
@laojianzi Please also fix the conflicts?

**Comment by thinkerou** [comment]
> @laojianzi Please also fix the conflicts?

we should remove `.travis.yml`?

**Comment by laojianzi** [comment]
Thank

@appleboy add go version 1.16 to ci and fixed the conflicts
@thinkerou I didn't think it was necessary to keep .travis.yml after using github actions, so I deleted it

Please review again

**Comment by thinkerou** [comment]
> Thank
> 
> @appleboy add go version 1.16 to ci and fixed the conflicts
> @thinkerou I didn't think it was necessary to keep .travis.yml after using github actions, so I deleted it
> 
> Please review again

ok, thanks!
wait @javierprovecho solve this: `continuous-integration/travis-ci Expected — Waiting for status to be reported`

**Review by appleboy — APPROVED** [review]

**Comment by appleboy** [comment]
@thinkerou Move the PR to v1.8 milestone?

**Comment by thinkerou** [comment]
> @thinkerou Move the PR to v1.8 milestone?

OK, merge to marster before removing `continuous-integration/travis-ci Expected — Waiting for status to be reported`

**Comment by thinkerou** [comment]
@laojianzi please fix conflict and we will merge it, thanks!

**Comment by laojianzi** [comment]
@thinkerou @appleboy https://golang.org/pkg/net/#pkg-note-BUG

Should tests requiring fd be skipped when run in windows?

**Comment by appleboy** [comment]
@laojianzi We can skip the windows testing first and maybe create another PR to track the testing on Windows.

File: .github/workflows/gin.yml
Comment (appleboy) [inline .github/workflows/gin.yml:15]: remove windows-latest

**Review by appleboy — APPROVED** [review]

**Comment by appleboy** [comment]
Waiting for @thinkerou approval.

**Review by thinkerou — APPROVED** [review]
lgtm, thanks a lot!

**Pull request merged**
